src/v/kafka/server/replicated_partition.cc

```diff
@@ -314,9 +314,19 @@ ss::future<error_code> replicated_partition::validate_fetch_offset(
      * receive data.
      */
 
-    // calculate log end in kafka offset domain
-    const auto log_end = model::next_offset(
-      _translator->from_log_offset(_partition->dirty_offset()));
+    // Calculate log end in kafka offset domain
+    model::offset log_end;
+    if (_partition->dirty_offset() < _partition->start_offset()) {
+        // Empty raft log: do not try to translate dirty offset, instead
+        // report that the log ends one index ahead of the start offset.
+        log_end = model::next_offset(
```

Comment:
Hmm so why `next_offset`? If `log_end` is exclusive (as I presume is the case, judging by the fact that in the normal case it is dirty_offset + 1), then `log_end` should be equal to the start offset.

Reply:
Looking at this again, the only place we use `log_end` is in deciding whether to loop and wait for data to commit (if the offset is ahead of hwm but within dirty range).  So we don't need to calculate this at all.

I've changed `log_end` to be an optional, and in the empty log case we leave it nullopt and skip the loop that would rely on the value of log_end.